File: ruby/collatz-conjecture/collatz_conjecture.rb
# Collatz Conjecture
class CollatzConjecture
  def self.steps(start)
    raise ArgumentError if start < 1

    count = 0
    return count if start.eql?(1)

    loop do
      count += 1
      start = start.even? ? start / 2 : 3 * start + 1

      return count if start.eql?(1)
    end
  end
end

CollatzConjecture.steps(1)

File: haskell/collatz-conjecture/src/CollatzConjecture.hs
module CollatzConjecture
  ( collatz
  ) where

nextCollatz :: Integer -> Integer
nextCollatz n
  | even n = div n 2
  | otherwise = 3 * n + 1

collatzCount :: Integer -> Integer -> Integer
collatzCount 1 c = c
collatzCount n c = collatzCount (nextCollatz n) (c + 1)

collatz :: Integer -> Maybe Integer
collatz n
  | n < 1 = Nothing
  | otherwise = Just (collatzCount n 0)
